### DISCUSSION TITLE
Qt6?

### DISCUSSION THREAD

**Discussion by rcalixte** [ilius/ayandict#6] [General]
Have you thought about porting this application to Qt6? Support for Qt5 will be phased out by the end of 2025. There are newly released [Qt bindings for Go](https://github.com/mappu/miqt) with support for Qt5 and Qt6 and the interface is mostly similar to the previous Qt bindings. I [recently converted](https://github.com/rcalixte/jqview/commit/5e16f1f44eebd4c70f36c3e09a427c95ffb106ac) a much smaller application to Qt6 without much fuss. (I did have one issue with a goroutine thread panicking because of an issue with the window widget pointer but I was able to work around it.) The author @mappu has also been extremely responsive. It would be great to reestablish a community around Go and Qt development!

**Comment by ilius** [comment]
Yes, I saw miqt, but didn't look much into it.

I actually want FreeBSD support myself, which the current binding does not support. Besides, it's discontinued.

Thanks for reminding me.

**Reply by rcalixte** [reply]
> I actually want FreeBSD support myself

Given that it's a simpler build process (the initial build can take a while), I wonder if it would be easier to port to FreeBSD. I had a rather easy go (no pun intended) with the Linux arm64 build, no extra steps were required.

**Comment by ilius** [comment]
It doesn't compile on FreeBSD.
I tested a small package that compiles and works on Linux.

```
# github.com/mappu/miqt/qt
In file included from gen_qabstractanimation.cpp:11:
cgo-gcc-export-header-prolog:58:69: error: unknown type name '_Bool'
cgo-gcc-export-header-prolog:59:70: error: unknown type name '_Bool'
cgo-gcc-export-header-prolog:89:74: error: unknown type name '_Bool'
cgo-gcc-export-header-prolog:98:61: error: unknown type name '_Bool'
cgo-gcc-export-header-prolog:99:64: error: unknown type name '_Bool'
cgo-gcc-export-header-prolog:106:97: error: unknown type name '_Bool'
cgo-gcc-export-header-prolog:110:80: error: unknown type name '_Bool'
cgo-gcc-export-header-prolog:114:85: error: unknown type name '_Bool'
cgo-gcc-export-header-prolog:163:73: error: unknown type name '_Bool'
cgo-gcc-export-header-prolog:165:75: error: unknown type name '_Bool'
cgo-gcc-export-header-prolog:194:76: error: unknown type name '_Bool'
cgo-gcc-export-header-prolog:234:63: error: unknown type name '_Bool'
cgo-gcc-export-header-prolog:235:64: error: unknown type name '_Bool'
cgo-gcc-export-header-prolog:331:74: error: unknown type name '_Bool'
cgo-gcc-export-header-prolog:332:74: error: unknown type name '_Bool'
cgo-gcc-export-header-prolog:333:74: error: unknown type name '_Bool'
cgo-gcc-export-header-prolog:338:80: error: unknown type name '_Bool'
cgo-gcc-export-header-prolog:367:92: error: unknown type name '_Bool'
cgo-gcc-export-header-prolog:370:90: error: unknown type name '_Bool'
fatal error: too many errors emitted, stopping now [-ferror-limit=]
```

**Reply by ilius** [reply]
Same error with qt6

**Reply by rcalixte** [reply]
I had issues on Alpine too starting with this same file so I think there's an issue with the Cgo compilation outside of the supported/known to work architectures.

**Reply by mappu** [reply]
Hi - I had this error with all new versions of Clang, including on Windows MSYS2.

It can be solved by using `export CGO_CXXFLAGS='-D_Bool=bool'`.

Probably a more correct fix is something about `--std=c++??`, but I'm not exactly sure.

**Comment by ilius** [comment]
`v3` branch is ready to use. But since we can't cross-compile anymore, I need Github actions to build for Windows and Mac.

**Reply by rcalixte** [reply]
Have you tried something like [this](https://dev.to/kristoff/zig-makes-go-cross-compilation-just-work-29ho)? Admittedly, this is how I was also first introduced to Zig. There's also a talk on the subject as well but the article is easy enough.

**Reply by ilius** [reply]
Very interesting. But I still much prefer to do it with Github actions, because Qt is big (first-time native compile takes like 10 minutes) and Zig compiler is LLVM which makes it much slower.

GoldenDict, qBittorent and some other Qt C++ projects use Github actions to build for releases and we can use them as sample, so it shouldn't be difficult.

**Reply by rcalixte** [reply]
I'm planning on caching the build with GitHub Actions when I get around to it. I haven't tested it yet but I'm hoping that a persistent cache cuts down the time needed.

**Reply by ilius** [reply]
I didn't find any free ARM64 MacOS image for Github action. So I'm trying Zig build.

**Reply by rcalixte** [reply]
I would love to hear how it works for you.

**Reply by ilius** [reply]
Script: [build-mac-cross](https://github.com/ilius/ayandict/blob/v3/build-mac-cross)
I get these errors:

```
error: unable to find dynamic system library 'resolv' using strategy 'paths_first'. searched paths: none
error: unable to find dynamic system library 'Qt6Widgets' using strategy 'paths_first'. searched paths: none
error: unable to find dynamic system library 'Qt6Gui' using strategy 'paths_first'. searched paths: none
error: unable to find dynamic system library 'Qt6Core' using strategy 'paths_first'. searched paths: none
error: unable to find dynamic system library 'Qt6MultimediaWidgets' using strategy 'paths_first'. searched paths: none
error: unable to find dynamic system library 'Qt6Multimedia' using strategy 'paths_first'. searched paths: none
error: unable to find dynamic system library 'Qt6Network' using strategy 'paths_first'. searched paths: none
error: unable to find dynamic system library 'Qt6Widgets' using strategy 'paths_first'. searched paths: none
error: unable to find dynamic system library 'Qt6Gui' using strategy 'paths_first'. searched paths: none
error: unable to find dynamic system library 'Qt6Core' using strategy 'paths_first'. searched paths: none
```

**Reply by rcalixte** [reply]
You likely need to add the library lookup paths via `-L` and to include either some C++ library, either via `-lc++` or `-lstdc++`.

**Reply by rcalixte** [reply]
A verbose option would be helpful to see what paths the compilation commands are looking for too.

**Reply by rcalixte** [reply]
I'm playing with the script locally to see if I can get anywhere too. It would be great to get this working.

**Reply by rcalixte** [reply]
We may need to get the local environment setup for macOS, assuming the files are available for download. I don't know that linking against the GNU `libresolv.so` will yield a binary that can be used cross-platform but let's see if we can get a potentially useful binary first.

I'm looking here: https://github.com/apple-oss-distributions/libresolv

I suspect that the libraries that aren't found are because they are assuming the macOS path for the libraries and we're running on non-macOS hosts.

**Reply by rcalixte** [reply]
With the correct build paths, I get this now:
```
error: unable to find framework 'CoreFoundation'. searched paths: 
 /host/Frameworks/CoreFoundation.framework/CoreFoundation.tbd
 /host/Frameworks/CoreFoundation.framework/CoreFoundation.dylib
 /host/Frameworks/CoreFoundation.framework/CoreFoundation
error: unable to find framework 'Security'. searched paths: 
 /host/Frameworks/Security.framework/Security.tbd
 /host/Frameworks/Security.framework/Security.dylib
 /host/Frameworks/Security.framework/Security
```

**Reply by ilius** [reply]
I figured it won't be very easy lol

Maybe this helps? https://github.com/fjolnir/CoreFoundation-Lite-Linux

**Reply by mappu** [reply]
I think Zig can't do it alone yet - https://github.com/ziglang/zig/issues/1349 - it will work only with a copy of the macOS SDK. There is a copy of the SDK in https://github.com/tpoechtrager/osxcross along with a working clang configuration.

I have an example using this for Qt 5 and macOS x86_64 [here](https://github.com/mappu/miqt/blob/master/docker/macos-cross-x86_64-sdk14.5-go1.19-qt5.15-dynamic.Dockerfile), and a cross-compile builds fine on a normal Linux Github Actions runner. But, it needs some adaptation for Qt 6 and macOS ARM64.

**Reply by ilius** [reply]
`macos-14` image is M1, available since 2024-01-30! IDK why I didn't find [this](https://github.blog/changelog/2024-01-30-github-actions-introducing-the-new-m1-macos-runner-available-to-open-source/).

**Reply by ilius** [reply]
https://github.com/ilius/ayandict/actions/runs/18247455363/job/51957326293
